src/sage/rings/function_field/drinfeld_modules/finite_drinfeld_module.py

```diff
@@ -329,33 +378,147 @@ def frobenius_charpoly(self, var='X', algorithm='crystalline'):
+            sage: phi.frobenius_charpoly()(phi.frobenius_endomorphism()).ore_polynomial()
+            0
+
+        ::
+
+            sage: phi = DrinfeldModule(A, [K(zk), z^8, z^7, z^20])
+            sage: phi.frobenius_charpoly(algorithm='CSA')
+            X^3 + (z2*T^2 + z2*T + z2 + 1)*X^2 + (T^4 + (2*z2 + 1)*T^3 + (z2 + 2)*T^2 + (2*z2 + 1)*T + z2 + 2)*X + T^6 + 2*z2*T^3 + 2*z2 + 1
+            sage: phi.frobenius_charpoly(algorithm='crystalline')
+            X^3 + (z2*T^2 + z2*T + z2 + 1)*X^2 + (T^4 + (2*z2 + 1)*T^3 + (z2 + 2)*T^2 + (2*z2 + 1)*T + z2 + 2)*X + T^6 + 2*z2*T^3 + 2*z2 + 1
+            sage: phi.frobenius_charpoly(algorithm='gekeler')
+            X^3 + (z2*T^2 + z2*T + z2 + 1)*X^2 + (T^4 + (2*z2 + 1)*T^3 + (z2 + 2)*T^2 + (2*z2 + 1)*T + z2 + 2)*X + T^6 + 2*z2*T^3 + 2*z2 + 1
+            sage: phi.frobenius_charpoly(algorithm='motive')
+            X^3 + (z2*T^2 + z2*T + z2 + 1)*X^2 + (T^4 + (2*z2 + 1)*T^3 + (z2 + 2)*T^2 + (2*z2 + 1)*T + z2 + 2)*X + T^6 + 2*z2*T^3 + 2*z2 + 1
+            sage: phi.frobenius_charpoly()(phi.frobenius_endomorphism()).ore_polynomial()
+            0
+
+        Check that ``var`` inputs are taken into account for cached
+        characteristic polynomials::
+
+            sage: Fq = GF(2)
+            sage: A.<T> = Fq[]
+            sage: K.<z> = Fq.extension(2)
+            sage: phi = DrinfeldModule(A, [z, 0, 1])
+            sage: phi.frobenius_charpoly()
+            X^2 + X + T^2 + T + 1
+            sage: phi.frobenius_charpoly(var='Foo')
+            Foo^2 + Foo + T^2 + T + 1
+            sage: phi.frobenius_charpoly(var='Bar')
+            Bar^2 + Bar + T^2 + T + 1
         """
-        # Throw an error if the user asks for an unimplemented algorithm
-        # even if the char poly has already been computed
-        method_name = f'_frobenius_charpoly_{algorithm}'
-        if hasattr(self, method_name):
+        # If no algorithm is specified, return cached data (if it
+        # exists), or pick an algorithm:
+        if algorithm is None:
             if self._frobenius_charpoly is not None:
-                return self._frobenius_charpoly
-            self._frobenius_charpoly = getattr(self, method_name)(var)
-            return self._frobenius_charpoly
-        raise NotImplementedError(f'algorithm \"{algorithm}\" not implemented')
+                return self._frobenius_charpoly.change_variable_name(var)
+            else:
+                if self.rank() < self._base_degree_over_constants:
+                    algorithm = 'crystalline'
+                else:
+                    algorithm = 'CSA'
+        # If an algorithm is specified, do not use cached data, even
+        # if it is possible:
+        method_name = f'_frobenius_charpoly_{algorithm}'
+        if not hasattr(self, method_name):
+            raise NotImplementedError(f'algorithm "{algorithm}" not implemented')
+        self._frobenius_charpoly = getattr(self, method_name)(var)
+        return self._frobenius_charpoly
```

Comment:
```suggestion
        self._frobenius_charpoly = getattr(self, method_name)
        return self._frobenius_charpoly.change_variable_name(var)
```
To avoid problem if `var` is not a variable name.